src/transformStateWithClones.js

```diff
@@ -1,13 +1,48 @@
 'use strict';
-
 /**
  * @param {Object} state
  * @param {Object[]} actions
- *
- * @return {Object[]}
+ * @returns {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const nextState = { ...state };
+
+  for (const currentAction of actions) {
+    switch (currentAction.type) {
+      case 'addProperties':
+        const extraData = currentAction.extraData;
+
+        for (const key in extraData) {
+          nextState[key] = extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        const keysToRemove = currentAction.keysToRemove;
+
+        for (const key of keysToRemove) {
+          if (nextState.hasOwnProperty(key)) {
+            delete nextState[key];
+          }
+        }
+        break;
+
+      case 'clear':
+        for (const key in nextState) {
+          delete nextState[key];
+        }
+        break;
+
+      default:
+        break;
```

Comment:
It would be nice to show some kind of error if `action.type` is different than we expect. Maybe use `console.error` to do this.